### ISSUE TITLE
New/alternate syntax for milk presets?

### ISSUE THREAD

**Issue by mbellew** [projectM-visualizer/projectm#129]
I'm just curious if there is any interest in trying a new syntax for milkdrop presets.  I find the current syntax a little painful and started imagining a more user friendly format.  If there is interest I could think on it more and make a proposal...

**Comment by revmischa** [comment]
I have no doubt it's possible and not so hard. The multiline syntax is really terrible. 
The downside is, of course, losing compatibility with milkdrop. 
Some syntax to specify using GLSL (instead of the current shader language, HLSL) would be cool while you're at it.
Even better would be a (web-based?) interactive editor that generates the preset file from some input fields and widgets. Then it could be consistent and compatible with milkdrop. Combine with live preview and it'd be a game changer.

**Comment by mbellew** [comment]
Compatibility is important, I was thinking we'd provide a python script for two-way conversion.  I'll try to play with this over the holiday.

**Comment by struktured** [comment]
@mbellew you can also write any parser you want by implementing your own preset factory implementation. Python does seem like the right choice.

**Comment by mbellew** [comment]
step one (for the glsl part):  I wrote a tool (src//hlsl2glsl/h2g) to convert hlsl preset into a glsl preset.  I did this I would be something to test.  You can't load the new format yet, but progress.  Feedback welcome.

https://github.com/mbellew/projectm/tree/glsl

**Comment by revmischa** [comment]
@mbellew that’s awesome but projectM already transpiles HLSL to GLSL.....

**Comment by mbellew** [comment]
Yes, I reused that functionality to rewrite a .milk file with hlsl into a .milk file wth glsl.

**Comment by revmischa** [comment]
So this program converts a .milk file to a GLSL milk file to make it editable as a "new preset version" file?
The goal being to make the preset easier to edit?

**Comment by mbellew** [comment]
Step 2: now I can load glsl .milk presets.  It looks like this:

warp_lang=glsl_330
warp_1=`shader_body {
warp_2=`    ret = texture(sampler_main, uv).xyz;
warp_3=`    ret += texture(sampler_blur2, uv).xyz * vec3 (0.1);
warp_4=`    ret /= vec3 (1.1);
warp_5=`    ret -= vec3 (0.02);
warp_6=`}
comp_lang=glsl_330
comp_1=`shader_body {
comp_2=`    ret = texture(sampler_main, uv).xyz * vec3 (1) + GetBlur1(uv) * vec3 (3) + GetBlur2(uv) * vec3 (0) + GetBlur3(uv) * vec3 (0) + vec3 (0);
comp_3=`    ret += GetBlur2((uv - vec2 (0.5)) * vec2 (0.333) + vec2 (0.5));
comp_4=`    ret *= vec3 (1);
comp_5=`}

**Comment by G-9999** [comment]
@mbellew is it possible with what you've created to transform an HLSL .milk preset into (not a GLSL .milk preset) but instead into a GLSL shader itself? I'm looking for a way to take projkectM HLSL milk presets and import them into other apps that use GLSL shaders like ShaderToy.com 😅

**Comment by mbellew** [comment]
Wrong pull request, check out https://github.com/projectM-visualizer/projectm/pull/134

There I added code to convert an existing .milk file to use glsl see src/hlsl2glsl/h2g_main.cpp

For getting the actual compiled GLSL, you can output  `generator.GetResult()` near the end of 
`ShaderEngine::compilePresetShader()`

I'm not sure how useful that would be in other tools, there's the issue of setting up the predefined textures (the main source texture, and the noise textures).

**Comment by kblaschke** [comment]
If we really want to enable preset authors to add GLSL shaders, I'd rather use a different prefix for the shader code while keeping the original ones as-is, e.g. using `glsl_comp` and `glsl_warp`. Auto-converting the shaders will be quite hard, as I've seen plenty that use global variables in HLSL, which aren't supported by GLSL. Even SPIRV-Cross can't translate these shaders to GLSL or SPIR-V after preprocessing them to (totally valid) HLSL. One has to manually rewrite these shaders to only use local variables and parameters.

I think this will be the one big compatibility issue we'll not be able to solve properly without editing the original presets.
